lib/topology.mli

```diff
@@ -0,0 +1,160 @@
+  (** A NUMA node providing CPU and memory resources *)
+
+  val v : affinity:CPUSet.t -> memfree:int64 -> t
+  (** [v ~affinity ~memfree] constructs a resource requiring
+   * affinity to be non-empty and memfree to be > 0.
+   * A zero request is allowed due to [shrink].
+   * *)
+
+  val empty : t
+  (** [empty] is a resource providing nothing *)
+
+  val union : t -> t -> t
+  (** [union a b] is a resource providing both [a] and [b] *)
+
+  val min_memory : t -> t -> t
+  (** [min_memory a b] is [a] with the memory set to the minimum of [a] and [b] *)
+
+  val pp_dump : t Fmt.t
+  (** [pp_dump ppf v] pretty-prints [v] on [ppf] *)
+end
+
+module NUMARequest : sig
+  type t = private {memory: int64; vcpus: int}
+  (** A (VM) requesting resources *)
+
+  val v : memory:int64 -> vcpus:int -> t
+  (**[v ~memory ~vcpus] constructs a request.
+   * [memory] and [vcpus] must be strictly positive. *)
+
+  val fits : t -> NUMAResource.t -> bool
+  (** [fits requested available] checks whether the [available] resources
+   * can satisty the [requested] resources. *)
+
+  val shrink : t -> NUMAResource.t -> t
+  (** [shrink a b] calculates the remaining resources that are needed,
+   * once all needed resources from [b] are allocated to [a].
+   * *)
+
+  val pp_dump : t Fmt.t
+  (** [pp_dump ppf v] pretty-prints [v] on [ppf] *)
+end
+
+module NUMA : sig
+  type t
+  (** Non Uniform Memory Access topology information: distances and CPUs.
+   * See ACPI version 6.2 section 5.2.17 "System Locality Distance Information Table (SLIT)".
+   * *)
+
+  (** A NUMA node index.
+   * Distinct from an int to avoid mixing with CPU numbers
+   * *)
+  type node = private Node of int
+
+  val v : distances:int array array -> cpu_to_node:int array -> t
+  (** [v distances cpu_to_node] stores the topology.
+   * [distances] is a square matrix [d] where [d.(i).(j)] is an approximation
+   * to how much slower it is to access memory from node [j] when running on node [i].
+   * Distances are normalized to 10, [d.(i).(i)] must equal to 10,
+   * and all values must be >= 10.
+   * Usually distances are symmetric [d.(i).(j) = d.(j).(i)], but this is not required.
+   *
+   * Although the standard allows nodes to be unreachable, this module doesn't
+   * support it and raises an exception (value 0xFF = ~0U Xen = -1 bindings)
+   * It is assumed that all NUMA nodes can reach all other NUMA nodes,
+   * although the standard allows a value of 0xFF to mean no connectivity, this is not supported
+   * by this module.
+   * (The value would get translated from 0xFF to ~0U in Xen and then to -1 by the bindings).
+   *
+   * [cpu_to_nodes.(i)] = NUMA node of CPU [i]
+   *
+   * NUMA nodes without any CPUs are accepted (to handle hard affinities).
+   *
+   * Raises [invalid_arg] if above constraints are not met
+   *
+   * A typical matrix might look like this:
+   *  10 21
+   *  21 10
+   *
+   * A more complicated assymetric distance matrix:
```

Comment:
this doesn't look asymmetric?

Reply:
e.g. nodes 0,1,2, distance from 0 to 1 and 2 is 16, but distance from 1 to 2 is 22, so if you choose these 3 the speeds between the nodes will be quite different.
It is better to pick 0,1,4 where all the distances are 16.

Symmetric is probably the wrong term to use, but it is an unusual matrix.